core/src/main/java/org/elasticsearch/tribe/TribeService.java

```diff
@@ -462,13 +495,59 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
                     }
                 }
             }
+            clusterStateChanged |= updateCustoms(tasks, metaData);
+            if (clusterStateChanged) {
+                newState.blocks(blocks);
+                newState.metaData(metaData);
+                newState.routingTable(routingTable.build());
+            }
+            return clusterStateChanged;
+        }
 
-            if (!clusterStateChanged) {
-                return currentState;
-            } else {
-                return ClusterState.builder(currentState).incrementVersion().blocks(blocks).nodes(nodes).metaData(metaData)
-                        .routingTable(routingTable.build()).build();
+        private boolean updateCustoms(List<ClusterChangedEvent> tasks, MetaData.Builder metaData) {
+            boolean clusterStateChanged = false;
+            ClusterChangedEvent latestTask = tasks.get(tasks.size() - 1);
+            final ClusterState tribeClientState = latestTask.state();
+            Set<String> changedCustomMetaDataTypeSet = tasks.stream()
+                    .map(ClusterChangedEvent::changedCustomMetaDataSet)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toSet());
+            final List<Node> tribeClientNodes = TribeService.this.nodes;
+            Map<String, MetaData.Custom> mergedCustomMetaDataMap = mergeChangedCustomMetaData(changedCustomMetaDataTypeSet,
+                    customMetaDataType -> {
+                        List<MetaData.Custom> tribeCustomMetaDataList = new ArrayList<>(tribeClientNodes.size());
+                        for (Node tribeClientNode : tribeClientNodes) {
+                            String currentTribeName = TRIBE_NAME_SETTING.get(tribeClientNode.settings());
+                            final MetaData.Custom custom;
+                            if (currentTribeName.equals(tribeName)) {
```

Comment:
Answering here - I would prefer keeping the logic the same for all nodes i.e., any custom change is processed based on the latest cluster state of those nodes. No need to special case the current node as I don't see what it buys us but complexity?  without it it can be a simple stream